llvm/include/llvm/IR/PatternMatch.h

```diff
@@ -460,6 +460,39 @@ template <typename Predicate> struct apf_pred_ty : public Predicate {
 //
 ///////////////////////////////////////////////////////////////////////////////
 
+template <typename APTy> struct custom_checkfn {
+  function_ref<bool(const APTy &)> CheckFn;
+  bool isValue(const APTy &C) { return CheckFn(C); }
+};
+
+/// Match an integer or vector where CheckFn(ele) for each element is true.
+/// For vectors, poison elements are assumed to match.
+inline cst_pred_ty<custom_checkfn<APInt>>
+m_CheckedInt(function_ref<bool(const APInt &)> CheckFn) {
+  return cst_pred_ty<custom_checkfn<APInt>>{CheckFn};
+}
+
+inline api_pred_ty<custom_checkfn<APInt>>
+m_CheckedInt(const APInt *&V, function_ref<bool(const APInt &)> CheckFn) {
```

Comment:
Looking at this again just now, what's the point of these methods returning a `const APInt *`? In that case, isn't this limited to splats anyway and you could just as well use `m_APInt` + a check and save you the lambda?

I think this should be providing the result as a `Constant *`, which would support non-splat values matching the predicate...

Reply:
Missed this... yeah think you are right. Should be a constant.

Reply:
Interesting the norm case seems to be to always export these as `APInt`, if you looks at something like `m_Negative`/`m_NonNegative` its the same deal.

Any objection to me replacing all of these?

Reply:
Assuming there are not a lot of uses of them, replacing them should be fine. Something to be careful about is that the direct replacement here would be m_APIntAllowPoison, not m_APInt.

If there are many uses, it may make sense to retain them as a convenience (I don't think m_CheckedInt with APInt makes sense as a convenience though, it's strictly more complicated than just writing out the check).

Reply:
See: https://github.com/llvm/llvm-project/pull/91377 decided not to do the other. Didn't seem like much upside and made code less compact.